app/controllers/users.js

```diff
@@ -144,35 +165,107 @@ class UsersController extends BaseController {
   async edit(ws, data) {
     const {
       id: requestId,
-      user_edit: { login, current_password, new_password },
+      user_edit: {
+        login,
+        full_name,
+        email,
+        phone,
+        current_password,
+        new_password,
+      },
     } = data;
 
-    const currentUser = await User.findOne({ login });
+    const userId = this.sessionRepository.getSessionUserId(ws);
+    const currentUser = await User.findOne({ _id: userId });
     if (!currentUser) {
       throw new Error(ERROR_STATUES.USER_LOGIN_OR_PASS.message, {
         cause: ERROR_STATUES.USER_LOGIN_OR_PASS,
       });
     }
 
-    if (!(await currentUser.isValidPassword(current_password))) {
+    if (
+      current_password &&
+      !(await currentUser.isValidPassword(current_password))
+    ) {
       throw new Error(ERROR_STATUES.INCORRECT_CURRENT_PASSWORD.message, {
         cause: ERROR_STATUES.INCORRECT_CURRENT_PASSWORD,
       });
     }
 
-    const updateUser = new User({ login, password: new_password });
-    await updateUser.encryptAndSetPassword();
-    await User.updateOne(
-      { login },
-      {
-        $set: {
-          password_salt: updateUser.params.password_salt,
-          encrypted_password: updateUser.params.encrypted_password,
-          updated_at: new Date(),
-        },
+    let updateParam = { updated_at: new Date() };
+
+    if (new_password) {
+      const updateUser = new User({ password: new_password });
+      await updateUser.encryptAndSetPassword();
+
+      updateParam["password_salt"] = updateUser.params.password_salt;
+      updateParam["encrypted_password"] = updateUser.params.encrypted_password;
+    }
+
+    delete data.user_edit["new_password"];
+    delete data.user_edit["current_password"];
+
+    if (login) {
+      const isDuplicateUser = await User.findOne({ login });
+      if (isDuplicateUser) {
+        throw new Error(ERROR_STATUES.LOGIN_ALREADY_IN_USE.message, {
+          cause: ERROR_STATUES.LOGIN_ALREADY_IN_USE,
+        });
       }
+      updateParam["login"] = login;
+    }
+
+    if (full_name) {
+      updateParam["full_name"] = full_name;
+    }
+
+    if (email) {
+      const isDuplicateUser = await User.findOne({ email });
+      if (isDuplicateUser) {
+        throw new Error(ERROR_STATUES.EMAIL_ALREADY_IN_USE.message, {
+          cause: ERROR_STATUES.EMAIL_ALREADY_IN_USE,
+        });
+      }
+      updateParam["email"] = email;
+    }
+
+    if (phone) {
```

Comment:
it repeats same code again and again.
Let's add uniq DB indexes. And then just call `User.findOneAndUpdate(` so it will throw an error if a login/phone/email already exists